### ISSUE TITLE
if train_with_fp_copy flag is false, does quant_aware training still works?

### ISSUE THREAD

**Issue by brisker** [NervanaSystems/distiller#362]
if train_with_fp_copy flag is false, does quant_aware training still works? (In my try, the two cases's performance differs a lot)

**Comment by levzlotnik** [comment]
Hi @brisker ,

As you can see in `distiller/quantization/range_linear.py`:
https://github.com/NervanaSystems/distiller/blob/61fab0d05f2fb974f9a6504bc3bf8388743ba766/distiller/quantization/range_linear.py#L1282-L1291

This suggests that quantization-aware training only exists with a floating point copy of the weights (this is also true for DoReFa, WRPN and PACT quantizers).  
I am not quite sure what do you mean by training in the case without an fp32 copy, could you please elaborate?

Thanks,
Lev

**Comment by brisker** [comment]
@levzlotnik 
maybe I have some misunderstanding of this. Previously I thought maybe we can do quant_aware training without fp_copy in pytorch(since I see the flag here: https://github.com/NervanaSystems/distiller/blob/master/distiller/quantization/quantizer.py#L334)....

**Comment by levzlotnik** [comment]
@brisker 
Quantization aware training is back-propagating the gradients onto the fp32 weights (using Straight Through Estimator i.e. STE), and the re-quantizing the fp32 weights in the next batch onto the quantized weights, and during the forward pass you use the quantized weights. 

In any case - I don't think that applying backprop on the quantized weights "makes sense", because these weights have a uniform spacing between the values while the gradient vector will not have such a uniform spacing. In essence - the quantized weights and the gradient "don't speak the same language", in real hardware these quantized weights will be stored as `int8` values, while the distribution of the gradient vector will certainly be fp32.  
You can read more [here](https://openreview.net/pdf?id=rJxVxiiDoX).  

About the `Quantizer` flag - it's there because the `Quantizer` class handles the reformatting of the model by itself, meaning the `Quantizer` class creates the fp32 copies of the weights, when required.

Let me know if you have more questions.

Lev
